protocol/contracts/nexus/Shard.sol

```diff
@@ -0,0 +1,555 @@
+        bool allow
+    ) external onlyElevatedAccess {
+        /// @notice admin should check shardId does not exist
+        if (partner == ZERO_ADDRESS) { revert ZeroAddress(); }
+        // bool success;
+        if (allow) {
+            /*success =*/ allowedPartnerShardIds[partner].add(shardId);
+            partnerShards.add(shardId);
+        } else {
+            /*success =*/ allowedPartnerShardIds[partner].remove(shardId);
+            partnerShards.remove(shardId);
+        }
+        emit PartnerAllowedShardIdSet(partner, shardId, allow);
+    }
+
+    /*
+     * Set regular shard that templars can mint for themselves
+     * @param shardId Shard ID
+     * @param allow If shard should be added or removed
+     */
+    function setShardId(uint256 shardId, bool allow) external onlyElevatedAccess {
+        /// @notice possible that partners are allowed to mint regular shards too
+        if (allow) {
+            regularShards.add(shardId);
+        } else {
+            regularShards.remove(shardId);
+        }
+        emit RegularShardIdSet(shardId, allow);
+    }
+
+    function setShardIds(uint256[] memory shardIds, bool[] memory allows) external onlyElevatedAccess {
+        uint256 _length = shardIds.length;
+        if (_length != allows.length) {  revert InvalidParamLength(); }
+        bool allow;
+        uint256 shardId;
+        for (uint i; i < _length;) {
+            shardId = shardIds[i];
+            allow = allows[i];
+            if (allow) {
+                regularShards.add(shardId);
+            } else {
+                regularShards.remove(shardId);
+            }
+            emit RegularShardIdSet(shardId, allow);
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    /*
+     * Set the shard IDs partners can mint
+     * @param partner Address of the partner
+     * @param shardIds Shard IDs
+     * @param flags If the partner can mint shard
+     */
+    function setPartnerAllowedShardIds(
+        address partner,
+        uint256[] memory shardIds,
+        bool[] memory flags
+    ) external onlyElevatedAccess {
+         if (partner == ZERO_ADDRESS) { revert ZeroAddress(); }
+         uint256 _length = shardIds.length;
+         bool allowed;
+         uint256 shardId;
+         if (_length != flags.length) { revert InvalidParamLength(); }
+         for (uint i; i < _length;) {
+            allowed = flags[i];
+            shardId = shardIds[i];
+
+            EnumerableSet.UintSet storage partnerShardIds = allowedPartnerShardIds[partner];
+            if (allowed && !partnerShardIds.contains(shardId)) {
+                /// @dev ignoring return values if add failed so that we don't revert in the iteration. to allow completion
+                /* success = */ partnerShardIds.add(shardId);
+                /// also keep track of all partner shards, for convenience guard checks when user mints
+                if (!partnerShards.contains(shardId)) {
```

Comment:
No need for this `if` check, since `add()` will internally check if it's there or not. So i think this is double counting as it stands.

Same as below with `remove()` too

Reply:
this one I agree, because there's no loop